File: platforms/apple/Sources/BotaAppleSDK/Host/EncryptedUploadV2MaterialRegistry.swift
import CryptoKit
@preconcurrency import Foundation

struct EncryptedUploadV2TransferEvidence: Equatable, Sendable {
    let ciphertextLength: UInt64
    let ciphertextSHA256: Data
    let manifestLength: UInt16
    let manifestSHA256: Data
    let blockCount: UInt32

    init(
        ciphertextLength: UInt64,
        ciphertextSHA256: Data,
        manifestLength: UInt16,
        manifestSHA256: Data,
        blockCount: UInt32
    ) {
        self.ciphertextLength = ciphertextLength
        self.ciphertextSHA256 = ciphertextSHA256
        self.manifestLength = manifestLength
        self.manifestSHA256 = manifestSHA256
        self.blockCount = blockCount
    }
}

struct EncryptedUploadV2ManifestSubmission:
    CustomDebugStringConvertible, CustomStringConvertible, Equatable, Sendable
{
    let manifest: Data
    let evidence: EncryptedUploadV2TransferEvidence

    init(manifest: Data, evidence: EncryptedUploadV2TransferEvidence) {
        self.manifest = manifest
        self.evidence = evidence
    }

    var description: String { "EncryptedUploadV2ManifestSubmission(<redacted>)" }
    var debugDescription: String { description }
}

struct EncryptedUploadV2MaterialProvider:
    CustomDebugStringConvertible, CustomStringConvertible, Sendable
{
    typealias StagingRequestProvider = @Sendable (EncryptedUploadV2TransferEvidence) async throws -> URLRequest
    typealias ManifestSubmitter = @Sendable (EncryptedUploadV2ManifestSubmission) async throws -> Void
    typealias Finalizer = @Sendable (EncryptedUploadV2TransferEvidence) async throws -> Void
    typealias ReceiptProvider = @Sendable (EncryptedUploadV2TransferEvidence) async throws -> Data
    typealias CancellationHandler = @Sendable () async throws -> Void

    let authorization: Data
    private let stagingRequestProvider: StagingRequestProvider
    private let manifestSubmitter: ManifestSubmitter
    private let finalizer: Finalizer
    private let receiptProvider: ReceiptProvider
    private let cancellationHandler: CancellationHandler

    init(
        authorization: Data,
        stagingRequest: @escaping @Sendable (EncryptedUploadV2TransferEvidence) async throws -> URLRequest,
        submitManifest: @escaping @Sendable (EncryptedUploadV2ManifestSubmission) async throws -> Void,
        finalize: @escaping @Sendable (EncryptedUploadV2TransferEvidence) async throws -> Void,
        completionReceipt: @escaping @Sendable (EncryptedUploadV2TransferEvidence) async throws -> Data,
        cancel: @escaping @Sendable () async throws -> Void
    ) {
        self.authorization = authorization
        stagingRequestProvider = stagingRequest
        manifestSubmitter = submitManifest
        finalizer = finalize
        receiptProvider = completionReceipt
        cancellationHandler = cancel
    }

    var description: String { "EncryptedUploadV2MaterialProvider(<redacted>)" }
    var debugDescription: String { description }

    func stagingRequest(for evidence: EncryptedUploadV2TransferEvidence) async throws -> URLRequest {
        try await stagingRequestProvider(evidence)
    }

    func submit(_ submission: EncryptedUploadV2ManifestSubmission) async throws {
        try await manifestSubmitter(submission)
    }

    func finalize(_ evidence: EncryptedUploadV2TransferEvidence) async throws {
        try await finalizer(evidence)
    }

    func completionReceipt(_ evidence: EncryptedUploadV2TransferEvidence) async throws -> Data {
        try await receiptProvider(evidence)
    }

    func cancel() async throws {
        try await cancellationHandler()
    }
}

enum EncryptedUploadV2MaterialRegistryError: Error, Equatable, Sendable {
    case invalidMaterialID
    case duplicateMaterialID
    case missingMaterial
    case invalidAuthorization
    case invalidEvidence
    case invalidStagingRequest
    case invalidManifest
    case invalidReceipt
}

enum EncryptedUploadV2TerminalOutcome: CaseIterable, Equatable, Sendable {
    case completed
    case cancelled
    case failed
}

struct EncryptedUploadV2PreparedMaterial: Equatable, Sendable {
    let authorization: Data
    let authorizationSHA256: Data
}

struct EncryptedUploadV2AcceptedReceipt: Equatable, Sendable {
    let receipt: Data
    let receiptSHA256: Data
}

actor EncryptedUploadV2MaterialRegistry {
    private struct Entry: Sendable {
        let registrationID: UUID
        let provider: EncryptedUploadV2MaterialProvider
    }

    private static let authorizationByteCount = 408
    private static let manifestByteCount = 580
    private static let receiptByteCount = 336
    private static let digestByteCount = 32
    private static let maximumMaterialIDByteCount = 128

    private var providers: [String: Entry] = [:]

    func register(id: String, provider: EncryptedUploadV2MaterialProvider) throws {
        guard Self.isValidMaterialID(id) else {
            throw EncryptedUploadV2MaterialRegistryError.invalidMaterialID
        }
        guard provider.authorization.count == Self.authorizationByteCount else {
            throw EncryptedUploadV2MaterialRegistryError.invalidAuthorization
        }
        guard providers[id] == nil else {
            throw EncryptedUploadV2MaterialRegistryError.duplicateMaterialID
        }
        providers[id] = Entry(registrationID: UUID(), provider: provider)
    }

    func preparedMaterial(id: String) throws -> EncryptedUploadV2PreparedMaterial {
        let provider = try requiredEntry(id).provider
        return EncryptedUploadV2PreparedMaterial(
            authorization: provider.authorization,
            authorizationSHA256: Self.sha256(provider.authorization)
        )
    }

    func stagingRequest(
        id: String,
        evidence: EncryptedUploadV2TransferEvidence
    ) async throws -> URLRequest {
        try Self.validate(evidence)
        let entry = try requiredEntry(id)
        let request = try await entry.provider.stagingRequest(for: evidence)
        try requireCurrent(id: id, registrationID: entry.registrationID)
        guard request.url?.scheme?.lowercased() == "https",
              request.httpMethod?.uppercased() == "PUT",
              request.httpBody == nil,
              request.httpBodyStream == nil
        else {
            throw EncryptedUploadV2MaterialRegistryError.invalidStagingRequest
        }
        return request
    }

    func submitManifest(
        id: String,
        manifest: Data,
        evidence: EncryptedUploadV2TransferEvidence
    ) async throws {
        try Self.validate(evidence)
        guard manifest.count == Self.manifestByteCount,
              evidence.manifestLength == Self.manifestByteCount,
              Self.sha256(manifest) == evidence.manifestSHA256
        else {
            throw EncryptedUploadV2MaterialRegistryError.invalidManifest
        }
        let entry = try requiredEntry(id)
        try await entry.provider.submit(.init(manifest: manifest, evidence: evidence))
        try requireCurrent(id: id, registrationID: entry.registrationID)
    }

    func finalizeAndReceiveReceipt(
        id: String,
        evidence: EncryptedUploadV2TransferEvidence
    ) async throws -> EncryptedUploadV2AcceptedReceipt {
        try Self.validate(evidence)
        let entry = try requiredEntry(id)
        try await entry.provider.finalize(evidence)
        try requireCurrent(id: id, registrationID: entry.registrationID)
        let receipt = try await entry.provider.completionReceipt(evidence)
        try requireCurrent(id: id, registrationID: entry.registrationID)
        guard receipt.count == Self.receiptByteCount else {
            throw EncryptedUploadV2MaterialRegistryError.invalidReceipt
        }
        return EncryptedUploadV2AcceptedReceipt(
            receipt: receipt,
            receiptSHA256: Self.sha256(receipt)
        )
    }

    func terminate(id: String, outcome: EncryptedUploadV2TerminalOutcome) async throws {
        guard let entry = providers.removeValue(forKey: id) else { return }
        if outcome != .completed {
            try await entry.provider.cancel()
        }
    }

    func contains(id: String) -> Bool {
        providers[id] != nil
    }

    private func requiredEntry(_ id: String) throws -> Entry {
        guard let entry = providers[id] else {
            throw EncryptedUploadV2MaterialRegistryError.missingMaterial
        }
        return entry
    }

    private func requireCurrent(id: String, registrationID: UUID) throws {
        guard providers[id]?.registrationID == registrationID else {
            throw EncryptedUploadV2MaterialRegistryError.missingMaterial
        }
    }

    private static func validate(_ evidence: EncryptedUploadV2TransferEvidence) throws {
        guard evidence.ciphertextLength > 0,
              evidence.ciphertextSHA256.count == digestByteCount,
              evidence.manifestLength == manifestByteCount,
              evidence.manifestSHA256.count == digestByteCount,
              evidence.blockCount > 0
        else {
            throw EncryptedUploadV2MaterialRegistryError.invalidEvidence
        }
    }

    private static func isValidMaterialID(_ value: String) -> Bool {
        !value.isEmpty
            && value.utf8.count <= maximumMaterialIDByteCount
            && !value.contains(where: \Character.isWhitespace)
    }

    private static func sha256(_ data: Data) -> Data {
        Data(SHA256.hash(data: data))
    }
}

File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/host/EncryptedUploadV2MaterialRegistry.kt
package dev.bota.sdk.internal.host

import dev.bota.sdk.EncryptedUploadV2Material
import dev.bota.sdk.EncryptedUploadV2TransferEvidence
import java.security.MessageDigest
import java.util.UUID
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import okhttp3.Request

internal enum class EncryptedUploadV2TerminalOutcome {
    Completed,
    Cancelled,
    Failed,
}

internal class EncryptedUploadV2MaterialRegistryException(message: String) :
    IllegalArgumentException(message)

internal data class EncryptedUploadV2MaterialLease(internal val registrationId: UUID)

internal class EncryptedUploadV2PreparedMaterial(
    authorization: ByteArray,
    authorizationSha256: ByteArray,
    val lease: EncryptedUploadV2MaterialLease,
) {
    val authorization: ByteArray = authorization.copyOf()
    val authorizationSha256: ByteArray = authorizationSha256.copyOf()
}

internal class EncryptedUploadV2AcceptedReceipt(receipt: ByteArray, receiptSha256: ByteArray) {
    val receipt: ByteArray = receipt.copyOf()
    val receiptSha256: ByteArray = receiptSha256.copyOf()
}

internal class EncryptedUploadV2MaterialRegistry {
    private data class Entry(val registrationId: UUID, val material: EncryptedUploadV2Material)

    private val mutex = Mutex()
    private val entries = mutableMapOf<String, Entry>()

    suspend fun register(id: String, material: EncryptedUploadV2Material) {
        requireMaterialId(id)
        if (material.authorization.size != AuthorizationBytes) {
            throw EncryptedUploadV2MaterialRegistryException("authorization must be exactly $AuthorizationBytes bytes")
        }
        mutex.withLock {
            if (id in entries) throw EncryptedUploadV2MaterialRegistryException("material ID is already registered")
            entries[id] = Entry(UUID.randomUUID(), material)
        }
    }

    suspend fun preparedMaterial(id: String): EncryptedUploadV2PreparedMaterial {
        val entry = requiredEntry(id)
        return EncryptedUploadV2PreparedMaterial(
            entry.material.authorization,
            sha256(entry.material.authorization),
            EncryptedUploadV2MaterialLease(entry.registrationId),
        )
    }

    suspend fun stagingRequest(
        id: String,
        lease: EncryptedUploadV2MaterialLease,
        evidence: EncryptedUploadV2TransferEvidence,
    ): Request {
        validateEvidence(evidence)
        val entry = requiredEntry(id, lease)
        val request = entry.material.stagingRequest(evidence)
        requireCurrent(id, entry.registrationId)
        val contentLength = request.body?.contentLength() ?: 0
        if (request.url.scheme.lowercase() != "https" || request.method.uppercase() != "PUT" || contentLength != 0L) {
            throw EncryptedUploadV2MaterialRegistryException(
                "staging request must be an HTTPS PUT with an empty template body",
            )
        }
        return request
    }

    suspend fun submitManifest(
        id: String,
        lease: EncryptedUploadV2MaterialLease,
        manifest: ByteArray,
        evidence: EncryptedUploadV2TransferEvidence,
    ) {
        validateEvidence(evidence)
        if (manifest.size != ManifestBytes ||
            evidence.manifestLength.toInt() != ManifestBytes ||
            !sha256(manifest).contentEquals(evidence.manifestSha256)
        ) {
            throw EncryptedUploadV2MaterialRegistryException("manifest evidence is invalid")
        }
        val entry = requiredEntry(id, lease)
        entry.material.submitManifest(manifest.copyOf(), evidence)
        requireCurrent(id, entry.registrationId)
    }

    suspend fun finalizeAndReceiveReceipt(
        id: String,
        lease: EncryptedUploadV2MaterialLease,
        evidence: EncryptedUploadV2TransferEvidence,
    ): EncryptedUploadV2AcceptedReceipt {
        validateEvidence(evidence)
        val entry = requiredEntry(id, lease)
        entry.material.finalize(evidence)
        requireCurrent(id, entry.registrationId)
        val receipt = entry.material.completionReceipt(evidence)
        requireCurrent(id, entry.registrationId)
        if (receipt.size != ReceiptBytes) {
            throw EncryptedUploadV2MaterialRegistryException("receipt must be exactly $ReceiptBytes bytes")
        }
        return EncryptedUploadV2AcceptedReceipt(receipt, sha256(receipt))
    }

    suspend fun terminate(id: String, outcome: EncryptedUploadV2TerminalOutcome) {
        val entry = mutex.withLock { entries.remove(id) } ?: return
        if (outcome != EncryptedUploadV2TerminalOutcome.Completed) entry.material.cancelOnce()
    }

    suspend fun terminate(
        id: String,
        lease: EncryptedUploadV2MaterialLease,
        outcome: EncryptedUploadV2TerminalOutcome,
    ) {
        val entry = mutex.withLock {
            val value = entries[id]
                ?.takeIf { it.registrationId == lease.registrationId }
                ?: throw EncryptedUploadV2MaterialRegistryException("material lease is missing or stale")
            entries.remove(id)
            value
        }
        if (outcome != EncryptedUploadV2TerminalOutcome.Completed) entry.material.cancelOnce()
    }

    suspend fun contains(id: String): Boolean = mutex.withLock { id in entries }

    private suspend fun requiredEntry(id: String): Entry = mutex.withLock {
        entries[id] ?: throw EncryptedUploadV2MaterialRegistryException("material is not registered")
    }

    private suspend fun requiredEntry(id: String, lease: EncryptedUploadV2MaterialLease): Entry = mutex.withLock {
        entries[id]
            ?.takeIf { it.registrationId == lease.registrationId }
            ?: throw EncryptedUploadV2MaterialRegistryException("material lease is missing or stale")
    }

    private suspend fun requireCurrent(id: String, registrationId: UUID) {
        if (mutex.withLock { entries[id]?.registrationId } != registrationId) {
            throw EncryptedUploadV2MaterialRegistryException("material callback completed after terminal cleanup")
        }
    }

    private fun validateEvidence(value: EncryptedUploadV2TransferEvidence) {
        if (value.ciphertextLength == 0uL ||
            value.ciphertextSha256.size != DigestBytes ||
            value.manifestLength.toInt() != ManifestBytes ||
            value.manifestSha256.size != DigestBytes ||
            value.blockCount == 0u
        ) {
            throw EncryptedUploadV2MaterialRegistryException("transfer evidence is invalid")
        }
    }

    private fun requireMaterialId(value: String) {
        if (value.isEmpty() || value.encodeToByteArray().size > 128 || value.any(Char::isWhitespace)) {
            throw EncryptedUploadV2MaterialRegistryException("material ID is invalid")
        }
    }

    private fun sha256(value: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(value)

    private companion object {
        const val AuthorizationBytes = 408
        const val ManifestBytes = 580
        const val ReceiptBytes = 336
        const val DigestBytes = 32
    }
}
